napi-inl.h

```diff
@@ -3618,6 +3618,159 @@ inline void AsyncWorker::OnWorkComplete(
   delete self;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+// ThreadSafeFunction class
+////////////////////////////////////////////////////////////////////////////////
+
+// static
+template <typename DataType, typename Finalizer,
+          typename Context, typename ResourceString>
+inline ThreadSafeFunction ThreadSafeFunction::New(napi_env env,
+                                           const Function& callback,
+                                           const Object& resource,
+                                           ResourceString resourceName,
+                                           size_t maxQueueSize,
+                                           size_t initialThreadCount,
+                                           DataType* data,
+                                           Finalizer finalizeCallback,
+                                           Context* context) {
+  static_assert(details::can_make_string<ResourceString>::value
+      || std::is_convertible<ResourceString, napi_value>::value,
+      "Resource name should be string convertible type");
+
+  napi_threadsafe_function tsFunctionValue;
+  auto* finalizeData = new details::FinalizeData<DataType, Finalizer>({
+      finalizeCallback, context });
+  napi_status status = napi_create_threadsafe_function(env, callback, resource,
+      Value::From(env, resourceName), maxQueueSize, initialThreadCount, data,
+      details::FinalizeData<DataType, Finalizer, Context>::WrapperWithHint,
+      finalizeData, CallJS, &tsFunctionValue);
```

Comment:
Are these arguments in the right order?

```
napi_create_threadsafe_function(
  napi_env env,
  napi_value func,
  napi_value async_resource,
  napi_value async_resource_name,
  size_t max_queue_size,
  size_t initial_thread_count,
  void* thread_finalize_data,
  napi_finalize thread_finalize_cb,
  void* context,
  napi_threadsafe_function_call_js call_js_cb,
  napi_threadsafe_function* result);
```

I think there is a mixup after `initial_thread_count` (it looks like the thread finalize data and context are switched?)

Reply:
Disregard, took a closer look at how `details::FinalizeData` works